### PR TITLE
Simplify BlockBuilder

### PR DESCRIPTION
**Pull request by dain**
## Description

This PR simplifies `BlockBuilder` by removing all write methods from the 
interface. Each `BlockBuilder` implementation still has a `write` method, and
`Type` implementations simply downcast the expected implementation type.

This change is only possible with the removal of the wrapper `BlockBuilder`
types SingleArrayBlockWriter, SingleMapBlockWriter, and SingleRowBlockWriter.
These classes have been replaced with a single lambda `writeEntry` method 
on `ArrayBlockBuilder`, `MapBlockBuilder`, and `RowBlockBuilder`, that calls
back with the nested field builders. Since there is now a single `writeEntry`
method, there is no need for the `beginBlockEntry` and `closeEntry` methods.

To simplify the transition to the new `writeEntry` method, this PR includes
a new helper class for building array, map, and row values in functions. The
`ArrayValueBuilder`, `MapValueBuilder`, and `RowValueBuilder` interfaces have
a simple build value method to build a single value of related kind, and each
of these has a related _buffered_ class that can be used to share allocation
buffers across multiple value build calls.  This replaces the pattern of 
using a `PageBuilder` as a buffer in function implementations.

Additionally, `MapBlockBuilder` has been extended with a `strictNotDistinctFrom`
that builds the map hash table using `NOT DISTINCT FROM` semantics. This is
removes the need to use `TypedSet` to determine duplicates, and is much more
efficient.  This behavior is also supported in `BufferedMapValueBuilder` with
the `createBufferedDistinctStrict` static factory method.

## Added
                                                                                     
* `Int128BlockBuilder.writeInt128`
* `Fixed12BlockBuilder.writeFixed12`
* `Block.getSlice` with `SliceOutput` argument (to replace `Block.writeBytesTo`)
* `ArrayValueBuilder.buildArrayValue` and `BufferedArrayValueBuilder`
* `MapValueBuilder.buildMapValue` and `BufferedMapValueBuilder`
* `RowValueBuilder.buildRowValue` and `BufferedRowValueBuilder`
* `NOT DISTINCT FROM` support in `MapBlockBuilder` and `BufferedMapValueBuilder`
* `ArrayBlockBuilder.writeEntry`
* `MapBlockBuilder.writeEntry`
* `RowBlockBuilder.writeEntry`
* `LambdaMetafactoryGenerator` to simplify bytecode generation of lambda method calls 

## Removed
* `BlockBuilder.writeBytes`
* `BlockBuilder.writeShort`
* `BlockBuilder.writeInt`
* `BlockBuilder.writeLong`
* `BlockBuilder.beginBlockEntry`
* `BlockBuilder.closeEntry`
* `SingleArrayBlockWriter`
* `SingleMapBlockWriter`
* `SingleRowBlockWriter`

## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# SPI
* BlockBuilder has been redesigned to remove write methods from the interface. ({issue}`issuenumber`)
* Array, map, and row build has been changed to use a single `writeEntry`. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/block/BufferedRowValueBuilder.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        if (fieldBuilders.get(0).getPositionCount() + 1 > bufferSize) {
+            fieldBuilders = fieldBuilders.stream()
+                    .map(field -> field.newBlockBuilderLike(bufferSize, null))
+                    .toList();
+        }
+
+        int startSize = fieldBuilders.get(0).getPositionCount();
+
+        // build the map
+        builder.build(fieldBuilders);
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/block/BufferedRowValueBuilder.java:66]: Same as `BufferedMapValueBuilder`, we need to handle exceptions here

File: core/trino-main/src/main/java/io/trino/connector/system/RuleStatsSystemTable.java
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/connector/system/RuleStatsSystemTable.java:1]: Typo in commit description: "also make is difficult" -> "also make it difficult"

File: core/trino-main/src/main/java/io/trino/connector/system/RuleStatsSystemTable.java [hunk 2]
```diff
@@ -98,12 +99,13 @@ public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHand
...
 
-            BlockBuilder mapWriter = blockBuilders.get("time_distribution_percentiles").beginBlockEntry();
-            for (Map.Entry<Double, Double> percentile : stats.getTime().getPercentiles().entrySet()) {
-                DOUBLE.writeDouble(mapWriter, percentile.getKey());
-                DOUBLE.writeDouble(mapWriter, percentile.getValue());
-            }
-            blockBuilders.get("time_distribution_percentiles").closeEntry();
+            MapBlockBuilder blockBuilder = (MapBlockBuilder) blockBuilders.get("time_distribution_percentiles");
+            blockBuilder.buildEntry((keyBuilder, valueBuilder) -> {
+                for (Map.Entry<Double, Double> percentile : stats.getTime().getPercentiles().entrySet()) {
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/connector/system/RuleStatsSystemTable.java:104]:
```java
stats.getTime().getPercentiles().forEach((key, value) -> {
    DOUBLE.writeDouble(keyBuilder, key);
    DOUBLE.writeDouble(valueBuilder, value);
});

File: core/trino-spi/pom.xml [hunk 3]
```diff
@@ -254,6 +254,77 @@
                                     <old>interface io.trino.spi.ptf.TableFunctionDataProcessor</old>
                                     <justification>Relocate table function infrastructure</justification>
                                 </item>
+                                <item>
```
Comment (electrum) [inline core/trino-spi/pom.xml:257]: Add a commit before this to delete all the previous exclusions, since they're for the previous release

File: core/trino-spi/pom.xml [hunk 4]
```diff
@@ -254,6 +254,77 @@
...
+                                    <ignore>true</ignore>
```
Comment (electrum) [inline core/trino-spi/pom.xml:258]: We actually don't need `<ignore>true</ignore>` since that is set at the top level

File: lib/trino-orc/src/test/java/io/trino/orc/OrcTester.java [hunk 5]
```diff
@@ -760,32 +763,32 @@ else if (TIMESTAMP_TZ_MICROS.equals(type) || TIMESTAMP_TZ_NANOS.equals(type)) {
...
-                    BlockBuilder mapBlockBuilder = blockBuilder.beginBlockEntry();
-                    for (Entry<?, ?> entry : map.entrySet()) {
-                        writeValue(keyType, mapBlockBuilder, entry.getKey());
-                        writeValue(valueType, mapBlockBuilder, entry.getValue());
-                    }
-                    blockBuilder.closeEntry();
+                    Type keyType = mapType.getKeyType();
+                    Type valueType = mapType.getValueType();
+                    ((MapBlockBuilder) blockBuilder).buildEntry((keyBuilder, valueBuilder) -> {
+                        for (Entry<?, ?> entry : map.entrySet()) {
```
Comment (electrum) [inline lib/trino-orc/src/test/java/io/trino/orc/OrcTester.java:777]: Can use `map.forEach()` here

File: core/trino-spi/src/main/java/io/trino/spi/block/Block.java [hunk 6]
```diff
@@ -73,6 +74,14 @@ default Slice getSlice(int position, int offset, int length)
         throw new UnsupportedOperationException(getClass().getName());
     }
 
+    /**
+     * Gets a slice at {@code offset} in the value at {@code position} into the {@code output} slice output.
+     */
+    default void getSlice(int position, int offset, int length, SliceOutput output)
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/block/Block.java:80]: This might be better named `writeSliceTo`, as that matches `writeBytesTo` and makes the calling code easier to read. The `get` name makes it appear that the value is discarded. Here's a changed caller in the next commit with the current naming:
```diff
- block.writeBytesTo(position, 0, entryLength, out);
+ block.getSlice(position, 0, entryLength, out);
```
With the proposed name:
```diff
- block.writeBytesTo(position, 0, entryLength, out);
+ block.writeSliceTo(position, 0, entryLength, out);
```

File: core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java [hunk 7]
```diff
@@ -88,56 +89,58 @@ public static void output(ListaggAggregationState state, BlockBuilder out)
...
+                block.getSlice(position, 0, entryLength, out);
+                context.outputLength += entryLength;
+                context.emittedEntryCount++;
+
+                return true;
+            });
+
+            if (context.overflow) {
+                if (state.isOverflowError()) {
+                    throw new TrinoException(
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java:126]: Unrelated wrapping change

File: core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java [hunk 8]
```diff
@@ -147,6 +149,9 @@ else if (type == DictionaryId.class) {
                 else if (type == MapHashTables.class) {
                     retainedSize += ((MapHashTables) field.get(block)).getRetainedSizeInBytes();
                 }
+                else if (type.getEnclosingClass() == MapBlockBuilder.class) {
```
Comment (electrum) [inline core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java:152]: Is this change in the correct commit?

File: core/trino-spi/pom.xml [hunk 9]
```diff
@@ -325,6 +325,70 @@
                                     <old>method io.trino.spi.block.BlockBuilder io.trino.spi.type.RowType::createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int, int)</old>
                                     <new>method io.trino.spi.block.RowBlockBuilder io.trino.spi.type.RowType::createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int, int)</new>
                                 </item>
+                                <item>
+                                    <ignore>true</ignore>
```
Comment (electrum) [inline core/trino-spi/pom.xml:329]: `<ignore>true</ignore>` not needed

File: core/trino-spi/pom.xml [hunk 10]
```diff
@@ -389,6 +389,61 @@
                                     <old>method io.trino.spi.block.BlockBuilder io.trino.spi.type.VarcharType::createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int)</old>
                                     <new>method io.trino.spi.block.VariableWidthBlockBuilder io.trino.spi.type.VarcharType::createBlockBuilder(io.trino.spi.block.BlockBuilderStatus, int)</new>
                                 </item>
+                                <item>
+                                    <ignore>true</ignore>
```
Comment (electrum) [inline core/trino-spi/pom.xml:393]: `<ignore>true</ignore>` not needed. Same for all commits

**Review by electrum — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java (hunk 8 above)
Comment (dain) [reply core/trino-main/src/test/java/io/trino/block/AbstractTestBlock.java:152]: yep

**Comment by mosabua** [comment]
fyi @osscm  .. this one made it into 421

**Comment by nineinchnick** [comment]
The developer guide still references `beginBlockEntry` and needs to be updated https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/develop/connectors.md#connectorrecordsetprovider
